### DISCUSSION TITLE
Should poppler include QT by default?

### DISCUSSION THREAD

**Discussion by lukaso** [Homebrew/discussions#3428] [Writing Formulae/Casks]
In looking at `poppler` and the projects that depend on it there are some anomalies. 

There are 3 categories of packages:

- ones that use Gtk3 and therefore would not benefit from QT
- ones that are command line only or use a different widget library
- ones that should depend on `poppler-qt5` if they require QT at all

In fact, of all the projects that `depends_on "poppler"`, I couldn't find any that clearly require QT.

Therefore my questions are:

- Would it make sense to rename `poppler` to `poppler-qt6` and create a new formula `poppler` that comes without QT as a dependency?
- Alternatively, would it make sense to create a `poppler-noqt` formula?
- And finally, would it make sense to move all these projects onto the new "no QT" formula (however named)?

I'm happy to take on the task of creating the new formula and moving the formulas over (though might take a bit of time).

Here's my analysis:

These formulas depend on gtk3:

- evince
- pdfpc
- pqiv
- pympress

These formulas don't have any need for QT that I can see (command line; different widget library; all had no reference to QT in source code):

- diff-pdf
- elf
- exiftran
- mat2
- pdf2djvu
- pdf2svg
- pdfgrep
- pdfsandwich
- vips

These formulas appear to have a QT connection:

- pdftoipe - However, the code refers to QT_SELECT=5 in a number of places and best case it is therefore pulling in the wrong version of QT. https://github.com/otfried/ipe-tools/blob/eca6ecc5e8ed3ca87ba7797db5a4d643fd016fde/obs/debian/rules#L10
- nuxeo - appears to require QT5 if anything.

Is see this as a bit of continuing cleanup from the removal of options: Remove options:
https://github.com/Homebrew/homebrew-core/commit/2264def15a72f93c5001a917d7997d44b87c40dd

My selfish reasons are that I want to build Gimp from homebrew and don't want to include QT unnecessarily in the package.

**Comment by danielnachun** [comment]
I have wondered this myself as `qt` is an unwieldy dependency and has only gotten worse since we've added QtWebEngine support to all platforms on Qt5 and all platforms except macOS Catalina on Qt6. It's especially problematic for Linux HPC users like myself who don't use any packages that depend on the Poppler QT bindings.  It may be possible for us to split the package as you've suggested into something along the lines of:
>`poppler`
>`poppler-qt5` which would depend on `poppler`
>`poppler-qt6` which would depend on `poppler`

I will add to your list that there are two other packages `dspdfviewer` and `gdal` that use `poppler-qt5`.  `gdal` is also just a library that doesn't use Qt, while `dspdfviewer` does actually use Qt.  I also looked more carefully at `pdftoipe` and it does not itself actually use Qt at all.  So in total we have 2 formulae which actually use the Qt5 Poppler bindings, 0 packages using the Qt6 bindings, and 19 packages not using the Qt bindings at all.

@Homebrew/core what do we think about splitting `poppler` along these lines?  It does seem kind of unreasonable to foist Qt and its massive dependency tree as a dependency upon so many users for just 2 packages which are not terribly popular (`dspdfviewer` has 56 installs in 30 days while `nuxeo` has just 30).  By contrast, `poppler` has ~25,000 installs/month and `poppler-qt5` has ~14,000 installs/month.

**Reply by carlocab** [reply]
Just get rid of the `qt` dependency in `poppler`. If users need it with `qt` support, they can use `poppler-qt5`.

If they need Poppler with Qt6, they can help migrate `poppler-qt5` and its dependency tree to Qt6. Or host their own `poppler-qt6` in their own tap. If their `poppler-qt6` formula is sufficiently popular, it'll show up in our analytics and we can migrate it to `homebrew-core`.

**Reply by danielnachun** [reply]
That seems reasonable.  `poppler-qt5` will still need to be used by `dspdfviewer` - I double checked the `nuxeo` code and I can't find any references to QT there. `gdal` also needs to be switched from `poppler-qt5` to `poppler`.  Overall this seems to require a pretty minimal set of changes.  

If we eventually pick up more formulae that actually need Poppler with Qt 6 support, we'll have to re-evaluate adding it as a formula.  The only two packages Arch Linux has that currently depend on their `poppler-qt6` package are `texstudio` and `gimagereader-qt`, neither of which currently have formulae.  There are quite a few others that use their `poppler-qt5` package, but we don't have formulae for any of those either.

**Comment by danielnachun** [comment]
I've opened https://github.com/Homebrew/homebrew-core/pull/105996 to remove the `qt` dependency in `poppler`.

**Comment by sunderme** [comment]
texstudio needs poppler-qt6 and I specifically provided pull-requests to make it happen here.
Killing sane functionality without providing a future path is not nice.

**Reply by muzimuzhi** [reply]
To add more info:

I've tried with `brew tap-new` to provide a [`poppler-qt6` formula](https://github.com/muzimuzhi/homebrew-texstudio-beta/blob/master/Formula/poppler-qt6.rb) from my own tap for texstudio. It works, but in a limited scope.

For example, only bottles for `big_sur` and `x86_64_linux` are generated, because the [workflow file `tests.yml`](https://github.com/Homebrew/brew/blob/30fda551e80f11c69f91c7b324b7b16a9d059b8e/Library/Homebrew/dev-cmd/tap-new.rb#L66-L111) (generated by `brew tap-new`) is hard-coded to only run on `macos-latest` and `ubuntu-latest`, while the real [`tests.yml`](https://github.com/Homebrew/homebrew-core/blob/master/.github/workflows/tests.yml) used by [homebrew-core](https://github.com/Homebrew/homebrew-core) may run on more runners and does more sophisticated tests.

Would homebrew welcome a new formula `poppler-qt6`? It will also help in making versions of all `poppler*` formulae synced, see [`synced_versions_formulae.json`](https://github.com/Homebrew/homebrew-core/blob/master/synced_versions_formulae.json).

**Reply by danielnachun** [reply]
Homebrew is not a library manager so we probably wouldn't accept `poppler-qt6` as a formula on its own given that no Homebrew formulae use it.  Assuming `texstudio` is compliant with our formula requirements, it could be added as formula and then a `poppler-qt6` formula could be more acceptable as it would be satisfying a dependency for `texstudio`.

**Reply by danielnachun** [reply]
Also regarding what is built by the taps, a `big_sur` bottle will also work on `monterey`, so the only platforms missing are Catalina (whose support will end soon) and native support for ARM64 macOS.  I'm hoping GitHub gets ARM64 macOS support soon as Apple isn't making any new Intel Macs so the need for supporting that as a target is growing more urgent.

**Reply by muzimuzhi** [reply]
> Homebrew is not a library manager so we probably wouldn't accept `poppler-qt6` as a formula on its own given that no Homebrew formulae use it.

There's always some formulae which are not used by any other formulae. Why a not-used-by-any `poppler-qt6` is less acceptable than a such `texstudio`? Maybe because `poppler` is considered more low-level and closer to CLI users/developers (thus more likely to be depended on by other formulae), but `texstudio` closer to GUI users. OK, popularity seems to be a stronger (rejection) reason.

To the `texstudio` case, it's currently provided as a cask and I think that's the right place for a multi-platform GUI (La)TeX editor.

**Reply by fxcoudert** [reply]
As was said already:

> Homebrew is not a library manager

texstudio is an app. poppler-qt6 is a library.

**Reply by carlocab** [reply]
Sorry for breaking your usage of the formula.

However, `poppler` has many users, and `qt` is a gigantic dependency. It doesn't seem correct to me to require `qt` for every user who uses `poppler` to benefit the very small number of users who need Poppler built with Qt6.

If it turns out that our assessment that only a very small fraction of `poppler` users want it built with Qt6 is incorrect, then we can revisit this decision.

**Reply by sunderme** [reply]
Homebrew is preinstalled on the github osx runners which allows package buiding on osx which provides the binaries homebrew uses for its casks. One could be proud of that but that is apparently not the case.

The sad part is that there is barely any effort involved to remedy the situation by copying the old poppler formula to poppler-qt6.

And finally I am curious how porting to qt6 should happen when the dependencies are not provided for qt6, especially since qt6 is nominally the main qt version in homebrew.
